driver/src/contracts/base_contract.rs

```diff
@@ -22,26 +34,81 @@ impl BaseContract {
         let web3 = web3::Web3::new(transport);
 
         let json: serde_json::Value = serde_json::from_str(&contents)?;
-        let abi: String = json.get("abi").ok_or("No ABI for contract")?.to_string();
+        let abi_string = json.get("abi").ok_or("No ABI for contract")?.to_string();
+        let abi = abi_string.as_bytes();
 
         let decoded_address = hex::decode(&address[2..])?;
         let contract_address: Address = Address::from(&decoded_address[..]);
-        let contract = Contract::from_json(web3.eth(), contract_address, abi.as_bytes())?;
+        let contract = Contract::from_json(web3.eth(), contract_address, abi)?;
+
+        let network_id = env::var("NETWORK_ID")?.parse()?;
+        let private_key = H256::from_str(&env::var("PRIVATE_KEY")?)?;
 
         Ok(BaseContract {
             contract,
             web3,
             event_loop,
+            abi: ethabi::Contract::load(abi)?,
+            network_id,
+            public_key: address_from_private_key(&private_key)?,
+            private_key,
         })
     }
 
-    pub fn account_with_sufficient_balance(&self) -> Option<Address> {
-        let accounts: Vec<Address> = self.web3.eth().accounts().wait().ok()?;
-        accounts
-            .into_iter()
-            .find(|&acc| match self.web3.eth().balance(acc, None).wait() {
-                Ok(balance) => !balance.is_zero(),
-                Err(_) => false,
-            })
+    pub fn send_signed_transaction<P>(
+        &self,
+        function: &str,
+        params: P,
+        options: Options,
+    ) -> Result<H256>
+    where
+        P: Tokenize,
+    {
+        let nonce = if let Some(nonce) = options.nonce {
+            nonce
+        } else {
+            self.web3
+                .eth()
+                .transaction_count(self.public_key, None)
+                .wait()?
+        };
+
+        info!("Sending tx from {} with nonce {:x}", self.public_key, nonce);
+
+        let signed_tx = self
+            .abi
+            .function(function)
+            .and_then(|function| function.encode_input(&params.into_tokens()))
+            .map(move |data| {
```

Comment:
Can you elaborate on the requirement for "move" here? I guess we are "forcing our closure to take ownership of its environment", but I don't really know

Reply:
You are right that it's not needed since none of the variables used are referenced outside the closure later. Thus a mutable borrow is fine.